Archive swap command

D4nte opened the pull request:
New command to archive swaps: `nectar archive-swap <swap id>`.
The command should be done while nectar is stopped as it updates the database.
Archiving a swap stops nectar to resume its automated execution, `create-transaction` command can then be used to recover funds.

Commit:
Import anyhow::Result

D4nte commented:
@da-kami @thomaseizinger What did we say about active peers? As the automation is off, it means nectar will not remove the peer from active peers once the swap is manually finished.

The easiest way would be to remove the peer as active when archiving a swap. Are we happy with that?

thomaseizinger commented:
> The easiest way would be to remove the peer as active when archiving a swap. Are we happy with that?

I think that makes the most sense yes. Once it is archived, it should be fine for nectar to start trading again with someone.

thomaseizinger commented on nectar/src/database.rs:
```diff
@@ -171,8 +171,32 @@ impl Database {
+        let mut new_swap = stored_swap.clone();
+        new_swap.archived = Some(true);
+
+        let key = serialize(&swap_id).context("Could not serialize swap id for db storage")?;
```
Most of our error messages use the sentence pattern of "failed to XYZ". Not sure how fussed you are about consistency here but I mildly prefer the "failed to" spelling.

Additionally, we usually lowercase those messages.

thomaseizinger commented on nectar/src/database.rs:
```diff
@@ -467,4 +494,20 @@ mod tests {
+
+        assert!(res.is_err());
+
+        true
```
This could just be:

```suggestion
        res.is_err()
```

D4nte commented:
Should I test that a DB file created with a previous version of nectar can be loaded with this change?
Should it be something we add to the CI (e.g. committing a test file)?

thomaseizinger commented on nectar/src/database.rs:
```diff
@@ -145,6 +145,13 @@ impl Database {
+                        (
+                            Ok(_),
+                            Ok(Swap {
+                                archived: Some(true),
```
Thinking out load: This means `archived: None` is equivalent to `archived: Some(false)`. I guess that makes sense.

thomaseizinger replied:
I think this filtering would be clearer if you add another `.filter` where you test for the `archived` property explicitly because this whole match expression seems to be about deserializing and hence I wouldn't include this "business logic" here.

thomaseizinger commented on nectar/CHANGELOG.md:
```diff
@@ -7,6 +7,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+### Added
+
+-   New command to archive swaps: `nectar archive-swap <swap id>`.
+    The command should be done while nectar is stopped as it updates the database.
```
```suggestion
    The command should only be used while nectar is stopped as it updates the database.
```

thomaseizinger commented on nectar/CHANGELOG.md:
```diff
@@ -7,6 +7,17 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+
+### Changed
+
+-   **DB Change**: Add an "archived" field to stored swap data;
```
This is not a separate feature, hence I would mention it twice.

What do you think about including a general notice at the beginning of the changelog that says: "This release includes changes to the DB schema, we recommend to backup your data directory before upgrading"?

thomaseizinger reviewed (approved):
Thanks!

A note regarding the commits: I personally find it easier to work with if the last commit of a patch-series is the one that adds the new feature because it can then easily squash more commits into it.

For this particular PR, I would probably squash commits 2,4,5,6 together into one as they don't really add any value just by themselves.

thomaseizinger commented:
> Should I test that a DB file created with a previous version of nectar can be loaded with this change?
> Should it be something we add to the CI (e.g. committing a test file)?

I would probably just test that once locally. There isn't really a point in testing that continuously because it is not something that can easily break in the future right? Actually, I guess that is only true for migrations - they run only once and then never again ...

I wouldn't mind integration tests that test if we can load a database from a specific version, like:

```rust
#[test]
fn can_load_nectar_0_1_0_database_without_errors() {}

#[test]
fn can_load_nectar_0_2_0_database_without_errors() {}

// etc
```

Commit:
New command to archive swaps: `nectar archive-swap <swap id>`

The command should only be used while nectar is stopped as it updates
the database. Archiving a swap stops nectar to resume its automated
execution, `create-transaction` command can then be used to recover
funds.

Commit:
Convert error text to "failed to ..." format

Commit:
Separate deserialisation and conversion to SwapKind

Commit:
Combine assertion and returned value

thomaseizinger commented on nectar/src/database.rs:
```diff
@@ -142,14 +142,15 @@ impl Database {
                 }
                 Err(err) => Some(Err(err).context("failed to retrieve swaps from DB")),
             })
+            .map(|res| res.map(|(swap, swap_id)| SwapKind::from((swap, swap_id))))
```
Does clippy not suggest you to use a function reference here?

```suggestion
            .map(|res| res.map(SwapKind::from))
```

D4nte replied:
You can see the CI passed :)

D4nte replied:
Will fix with follow up PR

D4nte commented:
> > Should I test that a DB file created with a previous version of nectar can be loaded with this change?
> > Should it be something we add to the CI (e.g. committing a test file)?
> 
> I would probably just test that once locally. There isn't really a point in testing that continuously because it is not something that can easily break in the future right? Actually, I guess that is only true for migrations - they run only once and then never again ...
> 
> I wouldn't mind integration tests that test if we can load a database from a specific version, like:
> 
> ```rust
> #[test]
> fn can_load_nectar_0_1_0_database_without_errors() {}
> 
> #[test]
> fn can_load_nectar_0_2_0_database_without_errors() {}
> 
> // etc
> ```

Will add this in follow up PR as I am encountering problems: #3314.

D4nte commented:
bors r+
